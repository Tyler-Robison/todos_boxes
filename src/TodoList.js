import React, { useState } from "react";

import NewTodoForm from "./NewTodoForm";
import Todo from './Todo'
import { v4 as uuid } from 'uuid'

const TodoList = () => {

    const startingTodos = [
        // { id: uuid() },
        // { id: uuid(), text: 'Take out trash' },
        // { id: uuid(), text: 'Make dinner' },
        // { id: uuid(), text: 'Feed cat' }
    ]

    const [todoList, setTodoList] = useState(startingTodos)

    // keep all todos currently in state, add {newTodo + id}
    const addTodo = (newTodo) => {
        setTodoList(todoList => [...todoList, { ...newTodo, id: uuid() }])
    }

    const deleteTodo = (id) => setTodoList(todoList.filter(todo => todo.id !== id))

    const editTodo = (id, formData) => {
        console.log('id', id)
        console.log('formData', formData)
        console.log('todolist', todoList)

        for (let i = 0; i < todoList.length; i++){
            if (todoList[i].id === id){
                todoList[i].text = formData.text
            }
        }
        setTodoList(todoList)
    }
    return (
        <div >
            <ol>
                {todoList.map(({ text, id }) => {
                    return <Todo
                        text={text}

                        id={id}
                        key={id}
                        deleteTodo={() => deleteTodo(id)} 
                        editTodo={() => editTodo(id)}/>
                })}
            </ol>
            <NewTodoForm addTodo={addTodo} />
        </div>
    )
}

export default TodoList;