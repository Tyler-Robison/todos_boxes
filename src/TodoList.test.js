import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from './TodoList'

function addTodo(todoList, text = 'test text') {

  const todoInput = todoList.getByLabelText('Todo Text')
  const btn = todoList.queryByText('Add Todo')

  fireEvent.change(todoInput, { target: { value: text } })
  fireEvent.click(btn)
}

it("renders without crashing", function () {
  render(<TodoList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it('should add a todo', function () {
  const todoList = render(<TodoList />)

  // haven't added yet
  expect(todoList.queryByText('Delete Todo')).not.toBeInTheDocument();

  addTodo(todoList);

  const removeButton = todoList.getByText('Delete Todo')
  expect(removeButton).toBeInTheDocument();
  expect(removeButton.previousSibling.previousSibling).toHaveTextContent('test text')

  // form should be empty
  expect(todoList.getAllByDisplayValue("")).toHaveLength(1);
})

it('should remove a todo', function () {
  const todoList = render(<TodoList />)

  expect(todoList.queryByText('Delete Todo')).not.toBeInTheDocument();
  addTodo(todoList);

  const removeButton = todoList.getByText('Delete Todo')
  expect(removeButton).toBeInTheDocument();
  fireEvent.click(removeButton)
  expect(removeButton).not.toBeInTheDocument();
})