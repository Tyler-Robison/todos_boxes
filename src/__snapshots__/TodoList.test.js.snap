// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <ol />
    <div>
      <form>
        <label
          for="text"
        >
          Todo Text
        </label>
        <input
          id="text"
          name="text"
          placeholder="text"
          type="text"
          value=""
        />
        <button>
          Add Todo
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
